### DISCUSSION TITLE
[Chapter 05] Marshalling ???

### DISCUSSION THREAD

**Discussion by ahngilwoong** [cowaine/spring-boot-for-msa#34] [Chapter 05. REST APIs with Spring MVC] [answered]
```
Q1. 마셜링은 Getter를 쓴다고 하는데 MappingJackson2HttpMessageConverter는 
    JsonData를 Binding할때(언마셜링?) Setter를 쓰나요 생성자를 쓰나요?
```

```
Q2. 스터디원분들이 생각하는 마셜링과 직렬화의 차이는 무엇인가요?
```

**Comment by JoisFe** [comment, marked as answer]
2번 질문 경우 제가한 질문과 같은데 #36 을 참고하시면 될 것 같습니다. @bunsung92 님 께서 감사하게도 답변을 해 주셨습니다.

1번 질문 경우 코드로 직접 느껴보면 좋을 것 같습니다.

- Json data -> 언마셜링 통해 HotelRoomReserveRequest에 바인딩
<img width="428" alt="image" src="https://github.com/cowaine/spring-boot-for-msa/assets/90208100/e28bf278-6629-4bfc-817a-af8f4f327690">

``` java
@RequestBody HotelRoomReserveRequest reserveRequest 
```

``` java
public class HotelRoomReserveRequest {

    private LocalDate checkInDate;

    private LocalDate checkOutDate;

    private String name;
}

```

1. Setter만 있는 경우
<img width="378" alt="image" src="https://github.com/cowaine/spring-boot-for-msa/assets/90208100/ae205245-0643-4954-848f-aa415c5f455b">

2. Getter만 있는 경우
<img width="460" alt="image" src="https://github.com/cowaine/spring-boot-for-msa/assets/90208100/d7404f49-7319-4bec-bcfe-5f6f42d42d0b">

- 둘다 데이터 바인딩이 잘 되는 것을 확인할 수 있습니다.
- 다만 여기에는 함정이... HotelRoomReserveRequest 에 아무생성자를 등록하지 않으면 기본생성자가 default로 존재합니다.
- 그렇다면 `@AllArgsConstructor를 통해 모든 필드를 초기화하는 생성자를 등록하면 (기본 생성자는 그럼 없어집니다.)

1. Setter + AllArgs
<img width="410" alt="image" src="https://github.com/cowaine/spring-boot-for-msa/assets/90208100/5434e6e5-e5dc-4794-af68-621ede615150">

2. Getter + AllArgs
<img width="410" alt="image" src="https://github.com/cowaine/spring-boot-for-msa/assets/90208100/e404da71-f1e6-46a8-89bc-894e3ed5b9db">

이 경우도 데이터 바인딩 잘 됩니다.

### 그럼 다른 방법 Getter, Setter 모두 없는 상황을 봅시다
1. NoArgs 만
<img width="587" alt="image" src="https://github.com/cowaine/spring-boot-for-msa/assets/90208100/51b021f9-7d69-410f-a18e-da03236c6573">

- 또잉 500 에러는 나지 않지만 데이터가 바인딩되지 않고 null이네요

2. AllArgs 만
<img width="420" alt="image" src="https://github.com/cowaine/spring-boot-for-msa/assets/90208100/fb62be4b-7139-424b-85f1-2bf7038019af">
- 이 친구는 또 데이터가 바인딩이 잘 되네요..

## 도대체 무슨 규칙이 ??
책에도 나와 있듯 Spring은 객체를 바인딩할 때 주로 JavaBean 규칙을 따라서 동작
즉 객체 필드에 접근이 가능해야합니다.
getter, setter를 통해 객체 필드에 접근이 가능하기에 데이터 바인딩을 할 필드를 식별할 수 있게 됩니다.

### 기본생성자가 있는 경우
1. Setter 가 있는 경우
- 기본생성자로 객체 초기화 후
- Setter 로 바인딩 할 필드 식별하고 값 설정
2. Getter 가 있는 경우
- 기본생성자로 객체 초기화 후
- Getter로 바인딩 할 필드 식별
- 그리고 Reflection을 이용해 값 설정

### 모든 필드 초기화하는 생성자가 있는 경우
- 뭐 당영히 되겠죵?

### 만약 특정 필드만 초기화하는 생성자가 있는 경우라면?
1. Getter, Setter 모두 없이 특정 필드만 초기화하는 생성자가 있는 경우
<img width="435" alt="image" src="https://github.com/cowaine/spring-boot-for-msa/assets/90208100/783d962a-8850-463f-80cb-a7eb4b041647">

- 위 경우 checkInDate, checkOutDate 2가지 필드만 초기화하는 생성자만 있는경우
- name만 null이고 나머지는 바인딩이 잘 됩니다. 또한 500에러도 나지 않습니다.

2. Getter + 특정 필드만 초기화하는 생성자가 있는 경우
<img width="541" alt="image" src="https://github.com/cowaine/spring-boot-for-msa/assets/90208100/034327d3-7c5a-4088-8563-b40557076bdd">

- 또잉... 모두 바인딩이 된다고??
- 지금까지 원리를 이해하면 당연히 2가지 필드는 바인딩 될테고
- 하나는 getter 메서드로 데이터 필드 식별 가능하며 Reflection 이용 ... 되겠군

3. Setter + 특정 필드만 초기화하는 생성자가 있는 경우
<img width="400" alt="image" src="https://github.com/cowaine/spring-boot-for-msa/assets/90208100/ab40e5e0-0504-4437-83bb-0aaae5aee561">

- 당연히 될 테고

### 500에러가 나는경우도 있음..
- name 만 초기화하는 생성자를 둔 경우

<img width="400" alt="image" src="https://github.com/cowaine/spring-boot-for-msa/assets/90208100/897779a9-2121-4b4d-83e0-8aa5a7e59d6d">
- 아니 이 경우도 위 예시를 보면 checkInDate, checkOutDate 가 null이고 name은 데이터 바인딩이 되어야하는거 아니야? 왜 500이...
- 이건 퀴즈!!
